### PR TITLE
Add mechanism to disable the "saved tensors hooks" feature

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #85972
* __->__ #85971

The rationale for this is that functorch doesn't work with saved
variable hooks at the moment or checkpointing and we need some way to
disable it.

Concretely:
- there's a context manager that does the disabling
- this feature is disabled on a thread-local basis
- one can set an error message or use the default error message that
says the feature has been disabled

Since it is thread local I needed to update ATen/ThreadLocalState. To
make things nicer, this PR refactors all the "saved tensors hooks"
related TLS things into a single struct.

Test Plan:
- new test

Differential Revision: [D39970936](https://our.internmc.facebook.com/intern/diff/D39970936)

### CODE REVIEW

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39970936).

File: test/test_autograd.py [hunk 1]
```diff
@@ -6718,6 +6718,32 @@ def test_setting_default_saved_variable_hooks_twice_should_use_inner(self):
...
+        with torch.autograd.graph.saved_tensors_hooks(lambda x: x, lambda x: x):
+            with self.assertRaisesRegex(RuntimeError, "error message"):
+                with torch.autograd.graph._disable_saved_tensors_hooks("error message"):
+                    pass
+
+        self.assertTrue(torch._C._autograd._saved_tensors_hooks_is_enabled())
+
+    def test_disabling_saved_tensor_hooks_nested(self):
+        with torch.autograd.graph._disable_saved_tensors_hooks("outer"):
+            with self.assertRaisesRegex(RuntimeError, "inner"):
```
Comment (albanD) [inline test/test_autograd.py:6738]: nit: htis context could be moved one line down right? After we disable with inner.

File: torch/autograd/graph.py [hunk 2]
```diff
@@ -132,3 +133,17 @@ def unpack_from_cpu(packed):
             return tensor.to(device, non_blocking=pin_memory)
 
         super().__init__(pack_to_cpu, unpack_from_cpu)
+
+
+@contextlib.contextmanager
+def _disable_saved_tensors_hooks(error_message):
```
Comment (albanD) [inline torch/autograd/graph.py:139]: Why not make this public?

**Review by albanD — APPROVED** [review]
Looks good!

**Review by soulitzer — APPROVED** [review]

File: test/test_autograd.py (hunk 1 above)
Comment (zou3519) [reply test/test_autograd.py:6738]: Yes

File: torch/autograd/graph.py (hunk 2 above)
Comment (zou3519) [reply torch/autograd/graph.py:139]: Do you want it as a public API?

**Comment by zou3519** [comment]
NB: this is a reland, the original PR got reverted

File: torch/autograd/graph.py (hunk 2 above)
Comment (albanD) [reply torch/autograd/graph.py:139]: If only functorch uses it, that's ok. But if more things starts to use it, I think we might as well make it a public API yes.

**Comment by zou3519** [comment]
I want to make sure this goes in before the branch cut, so I'll handle the follow-ups in a separate PR (it will probably take a non-zero amount of time b/c I'll have to figure out docs and the public API tests)

**Pull request closed without merging**